### ISSUE TITLE
Fails to apply index with no entries available to apply according to the reader

### ISSUE THREAD

**Issue by npepinpe** [zeebe-io/zeebe#3484]
**Describe the bug**

In a benchmark set up, a node was restarted due to OOM (other issue); on restart, one partition never managed to apply a single index (see `ZeebeRaftStateMachine`).

Current behaviour is that the Raft (see `PassiveRole`) will append entries on request, then compute the new commit index for the follower node, which is typically the minimum value of the last appended entry and the leader's commit index. It will the notify the state machine to apply all entries from the last commit index to this new commit index. This means that a follower's commit index is never greater than its last available (non-committed) entry index, meaning there is **always** an entry for a follower's commit index.

That said, in the benchmark we saw the error on restart that according to the state machine's log, its last index was much lower than the commit index (according to the `PassiveRole`), and that NO events were available to read (i.e. `reader.hasNext()` returned false):

```
[raft-server-2-raft-atomix-partition-8-state] ERROR io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Cannot apply index 28911258
[raft-server-2-raft-atomix-partition-8-state] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Expected next index is 25169449 and reader hasNext false
```

This is repeated, with the very first statement on restart being:

```
[raft-server-2-raft-atomix-partition-8-state] ERROR io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Cannot apply index 25169449
[raft-server-2-raft-atomix-partition-8-state] DEBUG io.zeebe.broker.clustering.atomix.ZeebeRaftStateMachine - Expected next index is 25169449 and reader hasNext false
```

The entry point here is when an `AppendRequest` is received by the follower, and like I mentioned, it will compute the index to ensure that we never try applying a non-existing index. Similarly, it will only call the state machine **after** the request entries have been successfully appended, and will **not** call it if it fails to append them.

**To Reproduce**

I will try to come up with a test case, but as of now I'm not sure how to reliably reproduce this.

**Expected behavior**

We should never try to apply non-existing entries.

**Comment by deepthidevaki** [comment]
I guess this is fixed by #4218  @npepinpe

**Comment by npepinpe** [comment]
Very likely, yes :tada:
